authentication app for user api

Zain-Bajwa opened the pull request:
### Testing Instructions
- Run `manage.py` file with command line `python manage.py runserver`
- You can hit user API with localhost by using postman:
  -  **Register a user**
http://127.0.0.1:8000/auth/register/
  - **Get a `JWT` token**
  http://127.0.0.1:8000/auth/token/create/
  - **Refresh a token**
  http://127.0.0.1:8000/auth/token/refresh/
  - **Get a user** by providing the user id and Bearer token in header with the method `GET`
 http://127.0.0.1:8000/auth/user/2/
  - **Update profile** by providing the user id and Bearer token in header with the method `PUT`

`Note`: All data is provided in `JSON` format.

Zain-Bajwa committed:
authentication app for user api

MAbdurrehman12 commented on authentication/views.py:
```diff
@@ -0,0 +1,79 @@
+    authentication_classes = [JWTAuthentication]
+    permission_classes = [IsAuthenticated,]
+    serializer_class = UserViewSerializer
+    lookup_field = 'pk'
```
Do we need a lookup_field in this viewset?

Zain-Bajwa replied:
- I have added __pycachy__ into .gitignore.
- For users registering only username, password and confirm_password are required. I have assigned the email as `null` by default.
- When the user enter incorrect username or password, it will return a message `Invalid username/password`.
- Also, resolve the issue for getting user by id.
- And yes, you are right we don't need to add lookup_field in `ListModelMixin` and `CreateModelMixin`.

MAbdurrehman12 commented on django_ecommerce/settings.py:
```diff
@@ -20,14 +21,15 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = 'django-insecure-g%2@m997h_a#=jtwvse5yrmae@3mb#5=fak-m93!=jmf6l$()k'
+SECRET_KEY = 'django-insecure-b=+&bq(#sm9srpii+13qo!7uq!ixxtg5s1h^m^o79efd54fqk0'
```
I think a better approach is to store this secret key in `.env` file and use it from there

Zain-Bajwa replied:
*Store secret key in `.env` file and also add `.env` file in `.gitignore`*

MAbdurrehman12 commented on django_ecommerce/settings.py:
```diff
@@ -20,14 +21,15 @@
+SECRET_KEY = 'django-insecure-b=+&bq(#sm9srpii+13qo!7uq!ixxtg5s1h^m^o79efd54fqk0'
 
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
```
Same as above. 
If this is stored in `.env`, we will have better flexibility for local and production. As on production, `DEBUG` must always be set to `False` as shown in the security warning message above.

Maybe we can do something like this: 

- Create `.env` file in `django_ecommerce`
- Put `.env` in gitignore (very important)
- Also create `.env.example` file with the env keys only (not the value) so other developers can get to know what env variables are being used inside the codebase.
- In `.env` file: set a variable `DEBUG_MODE=TRUE/FALSE`
- In `settings.py`: set `DEBUG = os.getenv("DEBUG_MODE") == "TRUE" `

MAbdurrehman12 reviewed (changes requested):
Overall, it looks good 👍 Just a couple of things I noticed: 

- I don't think we need to push cache files (`__pycache__`) too. We can add them into `.gitignore`. btw If you're not sure why these files are generated and what are the benefits then it will be worth searching and reading about it.

- Registering new users: 
    - Do we need both: email and username? I think one of them should be fine, usually email is preferred. To get rid of username, set `username = None` in your custom user model. but its not v important, maybe you can talk to @Wassaf-Shahzad if this is required. 
    - Password validation should also be done: 
        - Weak passwords should not be allowed, currently, I can even set "1" as my password.
        -  I think adding `Confirm Password` field will also be valuable.
    - I don't think its a good idea to send hashed/ecrypted password in response.

- Creating token:
    - If I enter a username which exists and enters a wrong password error is `No active account found with the given credentials` which can be misleading for the user. I think correct msg should be `Invalid password` or `Invalid email/password` whichever is easy

- Getting a user by ID: (/auth/user/2/)
  - If user is not found, Code breaks and 500 (server error) is sent in response. I think an appropriate msg should be sent in response (i.e: User not found etc).

Zain-Bajwa committed:
Set code format

Zain-Bajwa committed:
Add __pycache__ in .gitignare

Zain-Bajwa committed:
Rebase to main second commit

Zain-Bajwa committed:
Fix all issues of first review

Zain-Bajwa committed:
Change the message of invalid username

MAbdurrehman12 reviewed (approved):
Looks good 👍

Wassaf-Shahzad commented:
@Zain-Bajwa You can merge this PR

arslanashraf7 commented:
@Zain-Bajwa Nice to see that you are making good progress 👍 .

@Wassaf-Shahzad @MAbdurrehman12 Massive thanks to both of you guys for looking around and helping with the review and training 👍 

I didn't review the PR except for some overall architecture, I might be missing something but here are a couple of things that might be nice if we could include them:

1. It would be nice to integrate GitHub Actions to this repo with some basic checks e.g code quality e.g. `pylint, pycodestyle and/or black`.
2. Later on, we might want to add some test cases too using `pytest`, not that the whole project should be covered by the tests but maybe if we can write a few for authentication or products API so @Zain-Bajwa could learn that too.

Zain-Bajwa committed:
Final commit after CI

Create main.yml for applying actions pycodestyle, pylint and pytest

Apply Django test, pylint and pycodestyle test in CI

Solve the error in main.yml file

added secret key

Secret key in GitHub Secrets

MAbdurrehman12 reviewed (approved):
Nice work 👍
